samza-core/src/main/java/org/apache/samza/clustermanager/ContainerProcessManager.java

```diff
@@ -403,29 +412,23 @@ public void onResourcesCompleted(List<SamzaResourceStatus> resourceStatuses) {
 
   @Override
   public void onStreamProcessorLaunchSuccess(SamzaResource resource) {
-    String containerId = resource.getContainerId();
-    String containerHost = resource.getHost();
-
-    // 1. Obtain the processor ID for the pending container on this resource.
-    String processorId = getPendingProcessorId(containerId);
-    LOG.info("Successfully started Processor ID: {} on Container ID: {} on host: {}",
-        processorId, containerId, containerHost);
-
-    // 2. Remove the container from the pending buffer and add it to the running buffer. Additionally, update the
-    // job-health metric.
-    if (processorId != null) {
-      LOG.info("Moving Processor ID: {} on Container ID: {} on host: {} from pending to running state.",
-          processorId, containerId, containerHost);
-      state.pendingProcessors.remove(processorId);
-      state.runningProcessors.put(processorId, resource);
-      if (state.neededProcessors.decrementAndGet() == 0) {
-        state.jobHealthy.set(true);
-      }
-      containerManager.handleContainerLaunchSuccess(processorId, containerHost);
-    } else {
-      LOG.warn("Did not find a pending Processor ID for Container ID: {} on host: {}. " +
-          "Ignoring invalid/redundant notification.", containerId, containerHost);
+    // Scenario 1: processor belongs to current attempt of the job.
+    // This means, the current AM had placed a request for the processor
+    // and hence containerId should be found in the pendingProcessor map
+    if (state.pendingProcessors.containsValue(resource)) {
+      handleNewProcessorLaunchSuccess(resource);
+      return;
+    }
+    // Scenario 2: Due to AM HA, processor could belong to the previous attempt of the job.
+    // This means, the current AM did not place a request for the processor as it was already running.
+    // Hence it will be in the runningProcessors map and not in the pendingProcessor Map
+    if (jobConfig.getApplicationMasterHighAvailabilityEnabled() && state.runningProcessors.containsValue(resource)) {
+      handleRunningProcessorLaunchSuccess(resource);
+      return;
     }
+
+    LOG.warn("Did not find a pending Processor ID for Container ID: {} on host: {}. "
+        + "Ignoring invalid/redundant notification.", resource.getContainerId(), resource.getHost());
```

Comment:
would suggest do 
```
if ()
else if()
else 
```
instead as it eliminates the need to use multiple control statements (return) across the function to make it more readable. It becomes tedious especially if this function evolves and starts having updates to local state.